plugin/pathvirt/pathvirt.cpp

```diff
@@ -0,0 +1,586 @@
+    while (colon = strchr(element, ':')) {
+        /* if we are at the request index, return pointer to start of element */
+        if (curr_ind == i)
+            return element;
+
+        /* otherwise, advance pointer to next element and bump current index */
+        element = colon + 1;
+        curr_ind++;
+    }
+
+    /* last element */
+    if (curr_ind == i)
+        return element;
+
+    /* not found */
+    return NULL;
+}
+
+/*
+ * clgetsize - returns size of an element pointed to by @element in the
+ *             list
+ */
+static size_t
+clgetsize(const char *colonList, const char *element)
+{
+    /* either calculate the element's length, or call
+     * strlen if element was last one */
+    const char *colon = strchr(element, ':');
+    return colon ? colon - element : strlen(element);
+}
+
+static void
+errCheckGetRestartEnv(int ret)
+{
+    /* ret == -1 is fine; everything else is not */
+    if (ret < -1 /* RESTART_ENV_NOT_FOUND */) {
+        JASSERT(ret != RESTART_ENV_TOOLONG).Text("pathvirt: DMTCP_PATH_PREFIX exceeds "
+                "maximum size (10kb). Use a shorter environment variable "
+                "or increase MAX_ENV_VAR_SIZE and recompile.");
+
+        JASSERT(ret != RESTART_ENV_DMTCP_BUF_TOO_SMALL).Text("dmtcpplugin: DMTCP_PATH_PREFIX exceeds "
+                "dmtcp_get_restart_env()'s MAXSIZE. Use a shorter "
+                "environment variable or increase MAXSIZE and recompile.");
+
+        /* all other errors */
+        JASSERT(ret >= 0).Text("Fatal error retrieving DMTCP_PATH_PREFIX "
+                "environment variable.");
+    }
+}
+
+/* This function gets called on every restart. Since no user threads are
+ * active at that point, it's safe to access the shared data without
+ * locks.
+ */
+void
+pathvirtInitialize()
+{
+    char tmp[MAX_ENV_VAR_SIZE] = {0};
+
+    /* necessary since we don't know how many bytes dmtcp_get_restart_env
+       will write */
+    memset(newPathPrefixList, 0, sizeof(newPathPrefixList));
+    /* Try to get the value of ENV_DPP from new environment variables,
+     * passed in on restart */
+    int ret = dmtcp_get_restart_env(ENV_NEW_DPP, newPathPrefixList,
+                                    sizeof(newPathPrefixList) - 1);
+    errCheckGetRestartEnv(ret);
+
+    /* If the user had modified the original buffer prior to checkpointing,
+     * we use it now.
+     */
+    if (tmpBufferModified) {
+        snprintf(oldPathPrefixList, sizeof(oldPathPrefixList), "%s", tmpOldPathPrefixList);
+        tmpBufferModified = false;
+        memset(tmpOldPathPrefixList, 0, sizeof(tmpOldPathPrefixList));
+    }
+
+    /* If the user specified the DMTCP_ORIGINAL_PATH_PREFIX on restart,
+     * this will overrided previous calls to set the original buffer.
+     */
+    ret = dmtcp_get_restart_env(ENV_ORIG_DPP, tmp, sizeof(tmp) - 1);
+    errCheckGetRestartEnv(ret);
+    if (ret == RESTART_ENV_SUCCESS) {
+        memset(oldPathPrefixList, 0, sizeof(oldPathPrefixList));
```

Comment:
@rohgarg I don't think this `memset` is necessary because `snprintf` NULL terminates the string for you